ext/opcache/Optimizer/sccp.c

```diff
@@ -0,0 +1,1463 @@
+	if (type == _IS_BOOL) {
+		ZVAL_BOOL(result, Z_TYPE_P(op1) == IS_TRUE || Z_TYPE_P(op1) == IS_FALSE);
+	} else {
+		ZVAL_BOOL(result, Z_TYPE_P(op1) == type);
+	}
+}
+
+/* The functions chosen here are simple to implement and either likely to affect a branch,
+ * or just happened to be commonly used with constant operands in WP (need to test other
+ * applications as well, of course). */
+static inline int ct_eval_func_call(
+		zval *result, zend_string *name, uint32_t num_args, zval **args) {
+	if (zend_string_equals_literal(name, "chr")) {
+		zend_long c;
+		if (num_args != 1 || Z_TYPE_P(args[0]) != IS_LONG) {
+			return FAILURE;
+		}
+
+		c = Z_LVAL_P(args[0]) & 0xff;
+		ZVAL_INTERNED_STR(result, ZSTR_CHAR(c));
+		return SUCCESS;
+	} else if (zend_string_equals_literal(name, "in_array")) {
+		zval *val;
+		if (num_args != 2 || Z_TYPE_P(args[1]) != IS_ARRAY) {
+			return FAILURE;
+		}
+
+		ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(args[1]), val) {
+			if (fast_equal_check_function(val, args[0])) {
+				ZVAL_TRUE(result);
+				return SUCCESS;
+			}
+		} ZEND_HASH_FOREACH_END();
+		ZVAL_FALSE(result);
+		return SUCCESS;
+	} else if (zend_string_equals_literal(name, "strpos")) {
+		const char *found;
+		if (num_args != 2
+				|| Z_TYPE_P(args[0]) != IS_STRING
+				|| Z_TYPE_P(args[1]) != IS_STRING
+				|| !Z_STRLEN_P(args[1])) {
+			return FAILURE;
+		}
+
+		found = zend_memnstr(
+			Z_STRVAL_P(args[0]), Z_STRVAL_P(args[1]), Z_STRLEN_P(args[1]),
+			Z_STRVAL_P(args[0]) + Z_STRLEN_P(args[0]));
+		if (found) {
+			ZVAL_LONG(result, found - Z_STRVAL_P(args[0]));
+		} else {
+			ZVAL_FALSE(result);
+		}
+		return SUCCESS;
+	} else if (zend_string_equals_literal(name, "count")) {
+		if (num_args != 1 || Z_TYPE_P(args[0]) != IS_ARRAY) {
+			return FAILURE;
+		}
+
+		ZVAL_LONG(result, zend_hash_num_elements(Z_ARRVAL_P(args[0])));
+		return SUCCESS;
+	} else if (zend_string_equals_literal(name, "array_key_exists")) {
+		zval *value;
+		if (num_args != 2 || Z_TYPE_P(args[1]) != IS_ARRAY ||
+				(Z_TYPE_P(args[0]) != IS_LONG && Z_TYPE_P(args[0]) != IS_STRING
+				 && Z_TYPE_P(args[0]) != IS_NULL)) {
+			return FAILURE;
+		}
+
+		if (fetch_array_elem(&value, args[1], args[0])) {
+			return FAILURE;
+		}
+		ZVAL_BOOL(result, value != NULL);
+		return SUCCESS;
+	} else if (zend_string_equals_literal(name, "trim")
+			|| zend_string_equals_literal(name, "rtrim")
+			|| zend_string_equals_literal(name, "ltrim")) {
+		zend_string *str;
+		int mode;
+		if ((num_args < 1 || num_args > 2) || Z_TYPE_P(args[0]) != IS_STRING
+				|| (num_args == 2 && Z_TYPE_P(args[1]) != IS_STRING)) {
+			return FAILURE;
+		}
+
+		mode = ZSTR_VAL(name)[0] == 'l' ? 1 : ZSTR_VAL(name)[0] == 'r' ? 2 : 3;
+		if (num_args == 2) {
+			str = php_trim(Z_STR_P(args[0]), Z_STRVAL_P(args[1]), Z_STRLEN_P(args[1]), mode);
+		} else {
+			str = php_trim(Z_STR_P(args[0]), NULL, 0, mode);
+		}
+		ZVAL_STR(result, str);
+		return SUCCESS;
```

Comment:
Now that the general code for executing functions exists below, should these functions be switched to use it as well?